### PR TITLE
feat(agent-memory): opt-in write-time fact consolidation

### PR DESCRIPTION
**Pull request by KIvanow**
## Summary
Adds opt-in write-time fact consolidation to the @betterdb/agent-memory MemoryStore tier (TypeScript and Python). consolidateFacts() distills a set of memories into atomic, deduplicated facts and writes each as its own memory while keeping the source memories, so recall is preserved. This productizes the LongMemEval "facts" lever (previously benchmark-only code in packages/retrieval/eval/longmemeval/facts.ts) as a first-class, tested library feature. It is off by default and gated behind a constructor flag.

## Changes
- Add MemoryStore.consolidateFacts(options): selects source memories by scope/tags/olderThanSeconds/maxImportance, calls a caller-provided extractFacts LLM seam, reconciles the results, and writes dated fact memories additively (no source deletion, unlike the lossy consolidate()).
  - Add subject-keyed reconciliation (reconcile / applyOps): newest date wins, tombstone drops a subject, restatements dedupe. Each fact's asserted date is preserved by prefixing it into the written content.
  - Add the consolidation constructor flag (true or { enabled, factSource, factImportance }), default OFF. Calling consolidateFacts() while disabled throws a clear error. Prior fact memories are excluded from the candidate scan (-@source:{fact}) so a re-run never re-distills its own output.
  - New types: Fact, FactExtractor, ConsolidateFactsOptions/ConsolidateFactsResult, ConsolidationConfig; exported from the package index.
  - Python parity port (betterdb-agent-memory): reconcile_facts.py, consolidate_facts(), matching types and exports.
  - README API docs updated in both packages. 

## Checklist
- [x] Unit / integration tests added - new reconcileFacts and consolidateFacts suites in both languages
  (177 TS tests, 125 Python tests passing)
- [x] Docs added / updated - MemoryStore API sections in both READMEs
- [ ] [Roborev](https://www.roborev.io/) review passed — run `roborev review --branch` or `/roborev-review-branch` in Claude Code *(internal)*
- [ ] Competitive analysis done / discussed *(internal)*
- [ ] Blog post about it discussed *(internal)*

---

> [!NOTE]
> **Medium Risk**
> New Valkey write/delete path can change stored fact rows when enabled; behavior depends on caller `extractFacts` quality and date/subject reconciliation rules, though it is gated off by default and does not delete source memories.
> 
> **Overview**
> Adds **opt-in fact consolidation** to `MemoryStore` in TypeScript and Python: **`consolidateFacts` / `consolidate_facts`** selects source memories (scope, tags, age, importance), runs a caller-supplied **`extractFacts`** LLM seam, and writes **additive** fact memories while **leaving sources intact** (unlike lossy `consolidate`).
> 
> Facts are keyed by **`subject`**, reconciled with **newest `date` wins** and **tombstones**, and persisted with optional **`subject` / `date`** hash fields plus dated content prefixes. Runs are **stateful**: existing fact rows are loaded, compared for idempotent re-runs, superseded facts are **write-then-delete**, and duplicate subjects from races are cleaned up. The feature is **off by default** via a new **`consolidation`** constructor flag (`ConsolidationConfig`); calling it when disabled errors. Candidate scans **exclude** prior fact `source` values; consolidate filters gain **`includeSource`** for loading stored facts.
> 
> New **`reconcile` / `applyOps`** helpers and types (`Fact`, `ConsolidateFactsResult`, etc.) are exported; README API docs and unit tests cover reconciliation edge cases (bracketed dateless statements, tombstones, idempotency).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c83543e59acb31ca81f109d98dc9bcb77ea2cc25. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/agent-memory/src/MemoryStore.ts [hunk 1]
```diff
@@ -790,6 +833,185 @@ export class MemoryStore {
...
+      'content',
+      'subject',
+      'source',
+      'LIMIT',
+      '0',
+      String(CONSOLIDATE_SCAN_LIMIT),
+      'DIALECT',
+      '2',
+    );
+    return parseFtSearchResponse(raw).map((hit) => parseMemoryItem(this.name, hit));
```
[the remark this answers is not in the document]
Comment (KIvanow) [reply packages/agent-memory/src/MemoryStore.ts:1025]: False positive. valkey-search `RETURN` loads fields straight from the underlying hash document, so it returns non-indexed fields too. Verified on the shipping version (valkey 9.1.0, search module 66048): an index whose schema is only `content` + `source` still returns `subject` via `FT.SEARCH ... RETURN 3 content subject source`. `subject` is deliberately not in the schema because it is only ever returned, never filtered (unlike `source`, which backs the `@source` / `-@source` clauses); indexing it as a TAG would just bloat a high-cardinality index for no query benefit. Cross-run reconcile works as shipped.

**Comment by jamby77** [comment]
Two correctness edges worth fixing before this is relied on:

1. **`[date] statement` ambiguity** (`reconcileFacts.ts:20`): a dateless fact whose statement starts with a bracket (e.g. `"[Q3] revenue target is 5M"`) round-trips into a spurious dated fact (`date="Q3"`), and `isNewer`'s string compare then drops a genuinely newer dated fact. Fix: persist `date` in its own hash field and use it as the source of truth for datedness, instead of inferring from a leading bracket (content encoding can stay the same, so the diff logic is untouched).

2. **Consolidation race** (`MemoryStore.ts:923`): `existingBySubject` uses `Map.set`, so two concurrent `consolidateFacts` runs that each write a fact for the same subject leave a duplicate that's silently dropped from tracking and orphaned forever. Minimal fix: retract duplicate-subject facts on the next run (self-heal); stronger fix is a per-scope lock. Applies to the Python mirror too.

Nit: one-line conditional at `reconcileFacts.ts:44`.

**Comment by KIvanow** [comment]
> Two correctness edges worth fixing before this is relied on:
> 
> 1. **`[date] statement` ambiguity** (`reconcileFacts.ts:20`): a dateless fact whose statement starts with a bracket (e.g. `"[Q3] revenue target is 5M"`) round-trips into a spurious dated fact (`date="Q3"`), and `isNewer`'s string compare then drops a genuinely newer dated fact. Fix: persist `date` in its own hash field and use it as the source of truth for datedness, instead of inferring from a leading bracket (content encoding can stay the same, so the diff logic is untouched).
> 2. **Consolidation race** (`MemoryStore.ts:923`): `existingBySubject` uses `Map.set`, so two concurrent `consolidateFacts` runs that each write a fact for the same subject leave a duplicate that's silently dropped from tracking and orphaned forever. Minimal fix: retract duplicate-subject facts on the next run (self-heal); stronger fix is a per-scope lock. Applies to the Python mirror too.
> 
> Nit: one-line conditional at `reconcileFacts.ts:44`.

these should be fixed as well

**Review by jamby77 — APPROVED** [review]
Approved. Clean productization of the facts lever — recall-safe (sources kept, write-before-delete), idempotent re-runs, datedness read from the persisted `date` field rather than inferred from content brackets, and faithful TS↔Python parity with strong test coverage.

Non-blocking note: the duplicate-subject race-recovery path dedups `priorFacts` by last-in-FT.SEARCH-order while `existingBySubject` keeps the first-seen row as canonical, so a run over pre-existing duplicate rows could write the older-dated statement and delete the newer one. Self-heals next run and no source data is lost, but consider deduping by newest-date. CI red is the unrelated pre-existing cache-proposals-sqlite failure (this PR only touches agent-memory packages).
